src/lib/foreign-field.unit-test.ts

```diff
@@ -0,0 +1,186 @@
+import { ProvablePure } from '../snarky.js';
+import { Group } from './core.js';
+import { FieldVar } from './field.js';
+import { ForeignField, createForeignField, limbBits } from './foreign-field.js';
+import { Scalar as Fq, Group as G } from '../provable/curve-bigint.js';
+import { expect } from 'expect';
+import { createEquivalenceTesters, throwError } from './testing/equivalent.js';
+import { test, Random } from './testing/property.js';
+import { Provable } from './provable.js';
+import { ZkProgram } from './proof_system.js';
+import { Circuit, circuitMain } from './circuit.js';
+import { Scalar } from './scalar.js';
+
+// toy example - F_17
+
+class SmallField extends createForeignField(17n) {}
+let x = new SmallField(16);
+x.assertEquals(-1); // 16 = -1 (mod 17)
+x.mul(x).assertEquals(1); // 16 * 16 = 15 * 17 + 1 = 1 (mod 17)
+
+// invalid example - modulus too large
+
+expect(() => createForeignField(1n << 260n)).toThrow(
+  'modulus exceeds the max supported size'
+);
+
+// real example - foreign field arithmetic in the Pallas scalar field
+
+class ForeignScalar extends createForeignField(Fq.modulus) {}
+
+// types
+ForeignScalar satisfies ProvablePure<ForeignScalar>;
+
+// basic constructor / IO
+{
+  let s0 = 1n + ((1n + (1n << limbBits)) << limbBits);
+  let scalar = new ForeignScalar(s0);
+
+  expect(scalar.value).toEqual([0, FieldVar[1], FieldVar[1], FieldVar[1]]);
+  expect(scalar.toBigInt()).toEqual(s0);
+}
+
+test(Random.scalar, (x0, assert) => {
+  let x = new ForeignScalar(x0);
+  assert(x.toBigInt() === x0);
+  assert(x.isConstant());
+});
+
+// test equivalence of in-SNARK and out-of-SNARK operations
+
+let { equivalent1, equivalent2, equivalentBool2, equivalentVoid2 } =
+  createEquivalenceTesters(ForeignScalar, (x) => new ForeignScalar(x));
+
+// arithmetic
+equivalent2((x, y) => x.add(y), Fq.add, Random.scalar);
+equivalent1((x) => x.neg(), Fq.negate, Random.scalar);
+equivalent2((x, y) => x.sub(y), Fq.sub, Random.scalar);
+equivalent2((x, y) => x.mul(y), Fq.mul, Random.scalar);
+equivalent1(
+  (x) => x.inv(),
+  (x) => Fq.inverse(x) ?? throwError('division by 0'),
+  Random.scalar
+);
+
+// equality
+equivalentBool2(
+  (x, y) => x.equals(y),
+  (x, y) => x === y,
+  Random.scalar
+);
+equivalentVoid2(
```

Comment:
what's the `Void2` mean?

Reply:
that we are testing a function which takes two arguments (`2`) and returns nothing (`Void`)